Add acceleration and jerk limits to velocity task constraints

tumaanna opened the pull request:
- jerk and acceleration limits added to velocity tasks
- assume interpretation of optimized variable (joint velocities) as rad/s 
    - different from OpenSoT interpretation which is in rad/timestep
- tested by limiting the output trajectory acc and jerk (cmd diff and second diff) by setting past desired trajectory in xbot model state

tumaanna committed:
add acceleration limits to velocity task

tumaanna committed:
add python bindings to acc limits

tumaanna committed:
modify acc contraints to be for optimized vel in m/s
- assume optimized variable v is in m/s, not in m/timestep
- implement update method instead of setBounds

tumaanna committed:
fix errors and include in aggregated constraints resource

tumaanna committed:
add bindings
- add new constraints to the pyopensot module
- fix small issues

tumaanna committed:
fix acc limits and add jerk limits

tumaanna committed:
acc and jerk limits without logging

tumaanna committed:
remove debug data

tumaanna committed:
improve formatting

tumaanna committed:
format with clang-format

tumaanna committed:
update units

tumaanna committed:
update Jerk constraint arguments

Volksfest commented on src/constraints/velocity/AccelerationLimits.cpp:
```diff
@@ -0,0 +1,60 @@
+{
+    // the joint velocities are obtained from the robot model
+    // these could be either the measured values, or commanded values
+    _v = _robot.getJointVelocity();
```
According to https://github.com/ADVRHumanoids/xbot2_interface/blob/master/include/xbot2_interface/common/state_interface.h#L24 with https://github.com/ADVRHumanoids/xbot2_interface/blob/master/include/xbot2_interface/common/types.h#L33 and assuming what they mean in the long long text here https://eigen.tuxfamily.org/dox/classEigen_1_1Ref.html I think this is not needed.
"This", hereby is meant as local variable ```_v``` defined as a member variable. Maybe @chpohl can tell there more since he seems to be quite more knowdledged/experienced in cpp.

TL;DR just make it a local ref variable since the method also just retrieves a reference. Doing it like this (especially _v being a Eigen Value Type - not ref - you make a unnecessary copy)... Maybe I go too deep to optimization but this will be handled each cycle :sweat_smile:

tumaanna replied:
fixed

Volksfest commented on src/constraints/velocity/AccelerationLimits.cpp:
```diff
@@ -0,0 +1,60 @@
+    {
+        // Note: assume optimized joint velocity is in rad/second
+        // this interpertation is different from the rest of OpenSoT which assumes rad/timestep
+        _lowerBound[i] = (-1.0 * std::fabs(_qDDotLimit[i]) * _dT + _v[i]);
+        _upperBound[i] = (1.0 * std::fabs(_qDDotLimit[i]) * _dT + _v[i]);
```
_qDDotLimit[i] * _dT is constant for each cycle until somebody forcefully changes the limits. You can precompute this.

tumaanna replied:
fixed

Volksfest commented on src/constraints/velocity/AccelerationLimits.cpp:
```diff
@@ -0,0 +1,60 @@
+    return _dT;
+}
+
+Eigen::VectorXd AccelerationLimits::getAccelerationLimits()
+{
+    return _upperBound / _dT;
+}
```
Isn't that actually wrong? Shouldn't you just return _qDDotLimit?

tumaanna replied:
You're right, this is not correct. It was just copied from the original implementation with "per timestep in mind" and never used. Although I think it might make more sense to just return the bounds.

Volksfest commented on include/OpenSoT/constraints/velocity/AccelerationLimits.h:
```diff
@@ -0,0 +1,64 @@
+     * @return the system sample time in [s]
+     */
+    double getDT();
+    void setBoundScaling(const double boundScaling);
```
I don't think we need BoundScaling from the constraint itself (we can do it ourselves from outside)

tumaanna replied:
removed this everywhere.

Volksfest commented on src/constraints/velocity/JerkLimits.cpp:
```diff
@@ -0,0 +1,60 @@
+{
+    // the joint velocity and acceleration are obtained from the robot model
+    // these could be either the measured values, or commanded values
+    _v = _robot.getJointVelocity();
+    _a = _robot.getJointAcceleration();
```
As in the Acc Limits, I think this is not a good idea and an unnecessary copy

tumaanna replied:
fixed

Volksfest commented on src/constraints/velocity/JerkLimits.cpp:
```diff
@@ -0,0 +1,60 @@
+    {
+        // Note: assume optimized joint velocity is in rad/second
+        // this interpertation is different from the rest of OpenSoT which assumes rad/timestep
+        _lowerBound[i] = (-1.0 * std::fabs(_qDDDotLimit[i]) * _dT * _dT + _a[i] * _dT + _v[i]);
+        _upperBound[i] = (1.0 * std::fabs(_qDDDotLimit[i]) * _dT * _dT + _a[i] * _dT + _v[i]);
```
Here, I would go with Horner-Scheme but I think the compiler will already do that better anyway.

Thus, instead of $`\Delta t \Delta t \dddot{q} + \Delta t \ddot{q} + \dot{q}`$ this $`\Delta t (\Delta t \dddot{q} + \ddot{q}) + \dot{q}`$

tumaanna replied:
fixed

Volksfest commented on src/constraints/velocity/JerkLimits.cpp:
```diff
@@ -0,0 +1,60 @@
+    return _dT;
+}
+
+Eigen::VectorXd JerkLimits::getJerkLimits()
+{
+    return _upperBound / (_dT * _dT);
+}
```
As in the Acc limits

tumaanna replied:
modified to return the current bounds

Volksfest commented on src/constraints/velocity/JerkLimits.cpp:
```diff
@@ -0,0 +1,60 @@
+void JerkLimits::setBoundScaling(const double boundScaling)
+{
+    this->_boundScaling = boundScaling;
+}
```
As above... Do you actually use the scaling?

tumaanna replied:
removed it

Volksfest commented on src/tasks/velocity/CartesianAdmittance.cpp:
```diff
@@ -248,12 +248,6 @@ bool OpenSoT::tasks::velocity::CartesianAdmittance::setRawParams(const Eigen::Ve
-    {
-        throw std::invalid_argument("TBD lambda > 0");
-    }
-
```
Shouldn't that be already merged with the gain tuning :sweat_smile: ?

tumaanna replied:
gone.

Volksfest commented on bindings/python/constraints/velocity.hpp:
The format changes are a bit annoying but I see that they don't have a format check...

tumaanna replied:
left it for now

tumaanna committed:
fix getters and improve efficiency

tumaanna committed:
add missing bindings

Allow setting raw params with lambda neq 0

Revert "Allow setting raw params with lambda neq 0"

This reverts commit e18d09c236af51d97032340672e338d1ba1578e9.

Volksfest reviewed (approved):
In my taste I would remove the parenthesis in the assignment of the bounds but I don't mind.
